dataretrieval/nwis.py

```diff
@@ -447,21 +447,43 @@ def get_pmcodes(parameterCd='All', **kwargs):
     Parameters (Additional parameters, if supplied, will be used as query parameters).
     ----------
         parameterCd: string or listlike
+        parameterNm: string
     Returns:
         DataFrame containing the USGS parameter codes and Metadata as tuple
     """
-    payload = {'radio_pm_search' : 'pm_search',
-               'pm_group' : 'All+--+include+all+parameter+groups',
-               'pm_search' : parameterCd,
-               'casrn_search' : None,
-               'srsname_search' : None,
-               'show' :  ['parameter_group_nm', 'casrn', 'srsname','parameter_units', 'parameter_nm'],
-               'format' : 'rdb'}
+    url = 'https://help.waterdata.usgs.gov/code/parameter_cd_nm_query?'
     
-    payload.update(kwargs)
-    url = WATERDATA_URL + 'pmcodes/pmcodes'
+    if parameterCd is None and parameterNm is None:
+        raise TypeError('Query must specify a parameter code (parameterCd = ) or name (parameterNm = )')
+        
+    if parameterCd is not None and parameterNm is not None:
+        raise TypeError('Query must specify a parameter name or number, not both)')
+    
+    if parameterNm is None and parameterCd is not None: # querying based on a parameter code or list of codes
+        if isinstance(parameterCd, str): # when a single code is given
+            if parameterCd.lower() == "all": # if querying ALL a different url is needed
+                payload = {'fmt':'rdb', 'group_cd':'%'}
+                url = "https://help.waterdata.usgs.gov/code/parameter_cd_query?"
+            else: # this is for querying a single parameter
+                payload = {'parm_nm_cd':parameterCd,'fmt':'rdb'}  
+        if isinstance(parameterCd, list): # Querying with a list of parameters
+            l = []
+            for param in parameterCd:
+                payload = {'parm_nm_cd':param,'fmt':'rdb'}
+                response = query(url, payload)
+                if len(response.text.splitlines()) < 10: # empty query
+                     raise TypeError('One of the parameter codes used is not valid, please try a different value')
+                l.append(_read_rdb(response.text))
+            return pd.concat(l), _set_metadata(response)
+
+    if parameterNm is not None and parameterCd is None: # querying based on a parameter name
+        parameterNm ='%{0}%'.format(parameterNm) # update to include partial matches
+        payload = {'parm_nm_cd':parameterNm,'fmt':'rdb'}  
+            
     response = query(url, payload)
-    return _read_rdb(response.text), _set_metadata(response, **kwargs)
+    if len(response.text.splitlines()) < 10: # empty query
```

Comment:
Can you check a response code instead?

Reply:
I don't think we can, the url just returns the same header with no information when there is nothing to return.